numba/tests/test_function.py

```diff
@@ -0,0 +1,575 @@
+import types as pytypes
+from numba import njit, function, cfunc, types, int64, int32
+import ctypes
+
+from .support import TestCase
+
+
+def dump(foo):  # FOR DEBUGGING, TO BE REMOVED
+    foo_type = function.fromobject(foo)
+    foo_sig = foo_type.signature()
+    foo.compile(foo_sig)
+    print('{" LLVM IR OF "+foo.__name__+" ":*^70}')
+    print(foo.inspect_llvm(foo_sig.args))
+    print('{"":*^70}')
+
+
+# Decorators for tranforming a Python function to different kinds of functions:
+def pure_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    func.pyfunc = func
+    return func
+
+
+def cfunc_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    f = cfunc(sig)(func)
+    f.pyfunc = func
+    return f
+
+
+def njit_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    f = njit(func)
+    f.pyfunc = func
+    return f
+
+
+def njit2_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    f = njit(sig)(func)
+    f.pyfunc = func
+    return f
+
+
+def ctypes_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    cfunc = cfunc_func(func, sig)
+    addr = cfunc._wrapper_address
+    if sig == int64(int64):
+        f = ctypes.CFUNCTYPE(ctypes.c_int64)(addr)
+        f.pyfunc = func
+        return f
+    raise NotImplementedError(
+        f'ctypes decorator for {func} with signature {sig}')
```

Comment:
I think that all these should have `sig=None` as default. This is because some of them don't even use the signature and it should be asserted that the `signature` is indeed `None` in those cases, this to prevent mistakes in future use.

Reply:
The `sig` argument was introduced in order to have a family of decorators available for testing.
Currently, we only test functions with signature `int64(int64)` but for njit decorated first-class functions support other signatures need to be tested as well.
To resolve this, I suggest making the usage of `sig` argument explicit and eliminate it where possible.

Reply:
Fixed.